extensions/email/src/main/java/org/mule/extension/email/internal/commands/ListCommand.java

```diff
@@ -70,15 +75,29 @@
               .output(emailContent)
               .attributes(attributes)
               .build();
-          list.add(operationResult);
+          retrievedEmails.add(operationResult);
         }
       }
-      return list;
+      if (deleteAfterRetrieve) {
+        deleteRetrieved(connection, folderName, retrievedEmails);
+      }
+      return retrievedEmails;
     } catch (MessagingException me) {
       throw new EmailException("Error while retrieving emails: " + me.getMessage(), me);
     }
   }
 
+  private <T extends BaseEmailAttributes> void deleteRetrieved(MailboxConnection connection, String folderName,
+                                                               List<OperationResult<Object, T>> retrievedEmails) {
+    retrievedEmails.forEach(e -> {
+      int number = e.getAttributes()
+          .orElseThrow(() -> new EmailException("Could not find attributes in retrieved email"))
+          .getNumber();
+      setFlagCommand.setByNumber(connection, folderName, DELETED, number);
```

Comment:
no, as pop3 cannot delete a single email without losing consistency in the mailbox folder, we need to mark all the emails and then perform a deletion by closing the folder.